File: whatsapp-templates/php/whatsapp_templates.php
<?php
/**
 * SendSeven API - WhatsApp Templates Example (PHP)
 *
 * Demonstrates how to list and send WhatsApp template messages using the SendSeven API.
 * Features:
 * - List available templates
 * - Send template with text parameters
 * - Send template with header (image/document)
 * - Handle template categories (marketing, utility, authentication)
 * - Error handling for template not found, unapproved templates
 */

// Load environment variables from .env file (simple implementation)
function loadEnv(string $path): void {
    if (!file_exists($path)) {
        return;
    }
    $lines = file($path, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
    foreach ($lines as $line) {
        if (strpos(trim($line), '#') === 0) {
            continue;
        }
        if (strpos($line, '=') !== false) {
            list($name, $value) = explode('=', $line, 2);
            putenv(trim($name) . '=' . trim($value));
        }
    }
}

loadEnv(__DIR__ . '/.env');

// Configuration from environment
$API_TOKEN = getenv('SENDSEVEN_API_TOKEN');
$TENANT_ID = getenv('SENDSEVEN_TENANT_ID');
$API_URL = getenv('SENDSEVEN_API_URL') ?: 'https://api.sendseven.com/api/v1';
$CHANNEL_ID = getenv('CHANNEL_ID');
$CONTACT_ID = getenv('CONTACT_ID');

/**
 * Get common headers for API requests.
 *
 * @return array Headers array for cURL
 */
function getHeaders(): array {
    global $API_TOKEN, $TENANT_ID;
    return [
        'Authorization: Bearer ' . $API_TOKEN,
        'X-Tenant-ID: ' . $TENANT_ID,
        'Content-Type: application/json',
    ];
}

/**
 * Make a GET request to the API.
 *
 * @param string $endpoint API endpoint
 * @param array $params Query parameters
 * @return array Response data
 * @throws Exception If the API request fails
 */
function apiGet(string $endpoint, array $params = []): array {
    global $API_URL;

    $url = $API_URL . $endpoint;
    if (!empty($params)) {
        $url .= '?' . http_build_query($params);
    }

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => getHeaders(),
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    return json_decode($response, true);
}

/**
 * Make a POST request to the API.
 *
 * @param string $endpoint API endpoint
 * @param array $payload Request body
 * @return array Response data
 * @throws Exception If the API request fails
 */
function apiPost(string $endpoint, array $payload): array {
    global $API_URL;

    $url = $API_URL . $endpoint;

    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_POST => true,
        CURLOPT_POSTFIELDS => json_encode($payload),
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => getHeaders(),
    ]);

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $error = curl_error($ch);
    curl_close($ch);

    if ($error) {
        throw new Exception("cURL Error: $error");
    }

    if ($httpCode >= 400) {
        throw new Exception("API Error $httpCode: $response");
    }

    return json_decode($response, true);
}

/**
 * List available WhatsApp templates.
 *
 * @param string|null $category Filter by category (MARKETING, UTILITY, AUTHENTICATION)
 * @param string $status Filter by status (default: APPROVED)
 * @return array List of template objects
 * @throws Exception If the API request fails
 */
function listTemplates(?string $category = null, string $status = 'APPROVED'): array {
    $params = ['status' => $status];
    if ($category !== null) {
        $params['category'] = $category;
    }

    $data = apiGet('/whatsapp/templates', $params);
    return isset($data['items']) ? $data['items'] : $data;
}

/**
 * Send a WhatsApp template message.
 *
 * @param string $channelId The UUID of the WhatsApp channel
 * @param string $contactId The UUID of the contact to send to
 * @param string $templateName Name of the approved template
 * @param string $languageCode Language code (default: en)
 * @param array $components Template components with parameters
 * @return array The created message object
 * @throws Exception If the API request fails
 */
function sendTemplateMessage(
    string $channelId,
    string $contactId,
    string $templateName,
    string $languageCode = 'en',
    array $components = []
): array {
    $payload = [
        'channel_id' => $channelId,
        'contact_id' => $contactId,
        'template_name' => $templateName,
        'language_code' => $languageCode,
    ];

    if (!empty($components)) {
        $payload['components'] = $components;
    }

    return apiPost('/messages/send/template', $payload);
}

/**
 * Send a template message with text parameters in the body.
 *
 * @param string $channelId The UUID of the WhatsApp channel
 * @param string $contactId The UUID of the contact
 * @param string $templateName Name of the approved template
 * @param array $bodyParams List of text values for body placeholders
 * @param string $languageCode Language code (default: en)
 * @return array The created message object
 * @throws Exception If the API request fails
 */
function sendTemplateWithTextParams(
    string $channelId,
    string $contactId,
    string $templateName,
    array $bodyParams,
    string $languageCode = 'en'
): array {
    $components = [
        [
            'type' => 'body',
            'parameters' => array_map(function ($param) {
                return ['type' => 'text', 'text' => $param];
            }, $bodyParams),
        ],
    ];

    return sendTemplateMessage($channelId, $contactId, $templateName, $languageCode, $components);
}

/**
 * Send a template message with an image header.
 *
 * @param string $channelId The UUID of the WhatsApp channel
 * @param string $contactId The UUID of the contact
 * @param string $templateName Name of the approved template
 * @param string $imageUrl URL of the header image
 * @param array $bodyParams Optional list of text values for body placeholders
 * @param string $languageCode Language code (default: en)
 * @return array The created message object
 * @throws Exception If the API request fails
 */
function sendTemplateWithHeaderImage(
    string $channelId,
    string $contactId,
    string $templateName,
    string $imageUrl,
    array $bodyParams = [],
    string $languageCode = 'en'
): array {
    $components = [
        [
            'type' => 'header',
            'parameters' => [
                ['type' => 'image', 'image' => ['link' => $imageUrl]],
            ],
        ],
    ];

    if (!empty($bodyParams)) {
        $components[] = [
            'type' => 'body',
            'parameters' => array_map(function ($param) {
                return ['type' => 'text', 'text' => $param];
            }, $bodyParams),
        ];
    }

    return sendTemplateMessage($channelId, $contactId, $templateName, $languageCode, $components);
}

/**
 * Send a template message with a document header.
 *
 * @param string $channelId The UUID of the WhatsApp channel
 * @param string $contactId The UUID of the contact
 * @param string $templateName Name of the approved template
 * @param string $documentUrl URL of the document
 * @param string $filename Display filename for the document
 * @param array $bodyParams Optional list of text values for body placeholders
 * @param string $languageCode Language code (default: en)
 * @return array The created message object
 * @throws Exception If the API request fails
 */
function sendTemplateWithHeaderDocument(
    string $channelId,
    string $contactId,
    string $templateName,
    string $documentUrl,
    string $filename,
    array $bodyParams = [],
    string $languageCode = 'en'
): array {
    $components = [
        [
            'type' => 'header',
            'parameters' => [
                ['type' => 'document', 'document' => ['link' => $documentUrl, 'filename' => $filename]],
            ],
        ],
    ];

    if (!empty($bodyParams)) {
        $components[] = [
            'type' => 'body',
            'parameters' => array_map(function ($param) {
                return ['type' => 'text', 'text' => $param];
            }, $bodyParams),
        ];
    }

    return sendTemplateMessage($channelId, $contactId, $templateName, $languageCode, $components);
}

/**
 * Handle and display template-specific errors.
 *
 * @param Exception $error The exception object
 */
function handleTemplateError(Exception $error): void {
    $message = $error->getMessage();
    preg_match('/API Error (\d+)/', $message, $matches);
    $statusCode = isset($matches[1]) ? (int)$matches[1] : 0;

    if ($statusCode === 404) {
        echo "Template not found: $message\n";
        echo "Tip: Verify the template name exists and is approved\n";
    } elseif ($statusCode === 400) {
        if (stripos($message, 'not approved') !== false) {
            echo "Template not approved: $message\n";
            echo "Tip: Only APPROVED templates can be sent\n";
        } elseif (stripos($message, 'parameter') !== false) {
            echo "Parameter mismatch: $message\n";
            echo "Tip: Ensure the number of parameters matches the template\n";
        } else {
            echo "Bad request: $message\n";
        }
    } elseif ($statusCode === 401) {
        echo "Authentication failed: Check your API token\n";
    } elseif ($statusCode === 403) {
        echo "Permission denied: Token may lack required scopes\n";
    } else {
        echo "Error: $message\n";
    }
}

/**
 * Validate required configuration.
 *
 * @return bool True if all required variables are set
 */
function validateConfig(): bool {
    global $API_TOKEN, $TENANT_ID, $CHANNEL_ID, $CONTACT_ID;

    $missing = [];
    if (!$API_TOKEN) $missing[] = 'SENDSEVEN_API_TOKEN';
    if (!$TENANT_ID) $missing[] = 'SENDSEVEN_TENANT_ID';
    if (!$CHANNEL_ID) $missing[] = 'CHANNEL_ID';
    if (!$CONTACT_ID) $missing[] = 'CONTACT_ID';

    if (!empty($missing)) {
        echo "Error: Missing required environment variables:\n";
        foreach ($missing as $var) {
            echo "  - $var\n";
        }
        return false;
    }
    return true;
}

// Main execution
function main(): void {
    global $CHANNEL_ID, $CONTACT_ID;

    if (!validateConfig()) {
        exit(1);
    }

    // Example 1: List all approved templates
    echo str_repeat('=', 60) . "\n";
    echo "Listing approved WhatsApp templates...\n";
    echo str_repeat('=', 60) . "\n";

    try {
        $templates = listTemplates();
        if (empty($templates)) {
            echo "No approved templates found.\n";
            echo "Create templates in the WhatsApp Business Manager first.\n";
            return;
        }

        echo "Found " . count($templates) . " template(s):\n\n";
        foreach (array_slice($templates, 0, 5) as $template) {
            echo "  Name: {$template['name']}\n";
            echo "  Category: {$template['category']}\n";
            echo "  Language: {$template['language']}\n";
            echo "  Status: {$template['status']}\n";
            echo "\n";
        }
    } catch (Exception $e) {
        handleTemplateError($e);
        return;
    }

    // Example 2: List templates by category
    echo str_repeat('=', 60) . "\n";
    echo "Listing MARKETING templates...\n";
    echo str_repeat('=', 60) . "\n";

    try {
        $marketingTemplates = listTemplates('MARKETING');
        echo "Found " . count($marketingTemplates) . " marketing template(s)\n";
    } catch (Exception $e) {
        handleTemplateError($e);
    }

    // Example 3: Send a template with text parameters
    echo "\n" . str_repeat('=', 60) . "\n";
    echo "Sending template with text parameters...\n";
    echo str_repeat('=', 60) . "\n";

    try {
        $message = sendTemplateWithTextParams(
            $CHANNEL_ID,
            $CONTACT_ID,
            'order_confirmation',
            ['John Doe', 'ORD-12345'],
            'en'
        );

        echo "Template message sent successfully!\n";
        echo "  Message ID: {$message['id']}\n";
        echo "  Status: {$message['status']}\n";
    } catch (Exception $e) {
        handleTemplateError($e);
        echo "\nNote: Update template_name to match your approved template\n";
    }

    // Example 4: Send template with image header
    echo "\n" . str_repeat('=', 60) . "\n";
    echo "Sending template with image header...\n";
    echo str_repeat('=', 60) . "\n";

    try {
        $message = sendTemplateWithHeaderImage(
            $CHANNEL_ID,
            $CONTACT_ID,
            'promotion_with_image',
            'https://example.com/promo-image.jpg',
            ['Summer Sale', '50%'],
            'en'
        );

        echo "Template with image sent successfully!\n";
        echo "  Message ID: {$message['id']}\n";
    } catch (Exception $e) {
        handleTemplateError($e);
        echo "\nNote: Update template_name to match your approved template\n";
    }

    // Example 5: Send template with document header
    echo "\n" . str_repeat('=', 60) . "\n";
    echo "Sending template with document header...\n";
    echo str_repeat('=', 60) . "\n";

    try {
        $message = sendTemplateWithHeaderDocument(
            $CHANNEL_ID,
            $CONTACT_ID,
            'invoice_template',
            'https://example.com/invoice.pdf',
            'Invoice-2026-001.pdf',
            ['$199.99'],
            'en'
        );

        echo "Template with document sent successfully!\n";
        echo "  Message ID: {$message['id']}\n";
    } catch (Exception $e) {
        handleTemplateError($e);
        echo "\nNote: Update template_name to match your approved template\n";
    }
}

main();

File: whatsapp-templates/ruby/whatsapp_templates.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

##
# SendSeven API - WhatsApp Templates Example (Ruby)
#
# Demonstrates how to list and send WhatsApp template messages using the SendSeven API.
# Features:
# - List available templates
# - Send template with text parameters
# - Send template with header (image/document)
# - Handle template categories (marketing, utility, authentication)
# - Error handling for template not found, unapproved templates

require 'net/http'
require 'json'
require 'uri'

# Load .env file if it exists
def load_env_file
  env_file = File.join(__dir__, '.env')
  return unless File.exist?(env_file)

  File.readlines(env_file).each do |line|
    line = line.strip
    next if line.empty? || line.start_with?('#')

    key, value = line.split('=', 2)
    ENV[key.strip] ||= value.strip if key && value
  end
end

load_env_file

# Configuration from environment
API_TOKEN = ENV['SENDSEVEN_API_TOKEN']
TENANT_ID = ENV['SENDSEVEN_TENANT_ID']
API_URL = ENV['SENDSEVEN_API_URL'] || 'https://api.sendseven.com/api/v1'
CHANNEL_ID = ENV['CHANNEL_ID']
CONTACT_ID = ENV['CONTACT_ID']

##
# Get common headers for API requests.
#
# @return [Hash] Headers hash
def get_headers
  {
    'Authorization' => "Bearer #{API_TOKEN}",
    'X-Tenant-ID' => TENANT_ID,
    'Content-Type' => 'application/json'
  }
end

##
# Make a GET request to the API.
#
# @param endpoint [String] API endpoint
# @param params [Hash] Query parameters
# @return [Hash, Array] Response data
# @raise [RuntimeError] If the API request fails
def api_get(endpoint, params = {})
  uri = URI("#{API_URL}#{endpoint}")
  uri.query = URI.encode_www_form(params) unless params.empty?

  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = uri.scheme == 'https'

  request = Net::HTTP::Get.new(uri)
  get_headers.each { |key, value| request[key] = value }

  response = http.request(request)

  if response.code.to_i >= 400
    raise "API Error #{response.code}: #{response.body}"
  end

  JSON.parse(response.body)
end

##
# Make a POST request to the API.
#
# @param endpoint [String] API endpoint
# @param payload [Hash] Request body
# @return [Hash] Response data
# @raise [RuntimeError] If the API request fails
def api_post(endpoint, payload)
  uri = URI("#{API_URL}#{endpoint}")

  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = uri.scheme == 'https'

  request = Net::HTTP::Post.new(uri)
  get_headers.each { |key, value| request[key] = value }
  request.body = payload.to_json

  response = http.request(request)

  if response.code.to_i >= 400
    raise "API Error #{response.code}: #{response.body}"
  end

  JSON.parse(response.body)
end

##
# List available WhatsApp templates.
#
# @param category [String, nil] Filter by category (MARKETING, UTILITY, AUTHENTICATION)
# @param status [String] Filter by status (default: APPROVED)
# @return [Array] List of template objects
# @raise [RuntimeError] If the API request fails
def list_templates(category = nil, status = 'APPROVED')
  params = { status: status }
  params[:category] = category if category

  data = api_get('/whatsapp/templates', params)
  data.is_a?(Hash) && data['items'] ? data['items'] : data
end

##
# Send a WhatsApp template message.
#
# @param channel_id [String] The UUID of the WhatsApp channel
# @param contact_id [String] The UUID of the contact to send to
# @param template_name [String] Name of the approved template
# @param language_code [String] Language code (default: en)
# @param components [Array] Template components with parameters
# @return [Hash] The created message object
# @raise [RuntimeError] If the API request fails
def send_template_message(channel_id, contact_id, template_name, language_code = 'en', components = [])
  payload = {
    channel_id: channel_id,
    contact_id: contact_id,
    template_name: template_name,
    language_code: language_code
  }

  payload[:components] = components unless components.empty?

  api_post('/messages/send/template', payload)
end

##
# Send a template message with text parameters in the body.
#
# @param channel_id [String] The UUID of the WhatsApp channel
# @param contact_id [String] The UUID of the contact
# @param template_name [String] Name of the approved template
# @param body_params [Array<String>] List of text values for body placeholders
# @param language_code [String] Language code (default: en)
# @return [Hash] The created message object
# @raise [RuntimeError] If the API request fails
def send_template_with_text_params(channel_id, contact_id, template_name, body_params, language_code = 'en')
  components = [
    {
      type: 'body',
      parameters: body_params.map { |param| { type: 'text', text: param } }
    }
  ]

  send_template_message(channel_id, contact_id, template_name, language_code, components)
end

##
# Send a template message with an image header.
#
# @param channel_id [String] The UUID of the WhatsApp channel
# @param contact_id [String] The UUID of the contact
# @param template_name [String] Name of the approved template
# @param image_url [String] URL of the header image
# @param body_params [Array<String>] Optional list of text values for body placeholders
# @param language_code [String] Language code (default: en)
# @return [Hash] The created message object
# @raise [RuntimeError] If the API request fails
def send_template_with_header_image(channel_id, contact_id, template_name, image_url, body_params = [], language_code = 'en')
  components = [
    {
      type: 'header',
      parameters: [{ type: 'image', image: { link: image_url } }]
    }
  ]

  unless body_params.empty?
    components << {
      type: 'body',
      parameters: body_params.map { |param| { type: 'text', text: param } }
    }
  end

  send_template_message(channel_id, contact_id, template_name, language_code, components)
end

##
# Send a template message with a document header.
#
# @param channel_id [String] The UUID of the WhatsApp channel
# @param contact_id [String] The UUID of the contact
# @param template_name [String] Name of the approved template
# @param document_url [String] URL of the document
# @param filename [String] Display filename for the document
# @param body_params [Array<String>] Optional list of text values for body placeholders
# @param language_code [String] Language code (default: en)
# @return [Hash] The created message object
# @raise [RuntimeError] If the API request fails
def send_template_with_header_document(channel_id, contact_id, template_name, document_url, filename, body_params = [], language_code = 'en')
  components = [
    {
      type: 'header',
      parameters: [{ type: 'document', document: { link: document_url, filename: filename } }]
    }
  ]

  unless body_params.empty?
    components << {
      type: 'body',
      parameters: body_params.map { |param| { type: 'text', text: param } }
    }
  end

  send_template_message(channel_id, contact_id, template_name, language_code, components)
end

##
# Handle and display template-specific errors.
#
# @param error [StandardError] The error object
def handle_template_error(error)
  message = error.message
  status_code = message.match(/API Error (\d+)/)&.[](1)&.to_i || 0

  case status_code
  when 404
    puts "Template not found: #{message}"
    puts 'Tip: Verify the template name exists and is approved'
  when 400
    if message.downcase.include?('not approved')
      puts "Template not approved: #{message}"
      puts 'Tip: Only APPROVED templates can be sent'
    elsif message.downcase.include?('parameter')
      puts "Parameter mismatch: #{message}"
      puts 'Tip: Ensure the number of parameters matches the template'
    else
      puts "Bad request: #{message}"
    end
  when 401
    puts 'Authentication failed: Check your API token'
  when 403
    puts 'Permission denied: Token may lack required scopes'
  else
    puts "Error: #{message}"
  end
end

##
# Validate required configuration.
#
# @return [Boolean] True if all required variables are set
def validate_config
  missing = []
  missing << 'SENDSEVEN_API_TOKEN' if API_TOKEN.nil? || API_TOKEN.empty?
  missing << 'SENDSEVEN_TENANT_ID' if TENANT_ID.nil? || TENANT_ID.empty?
  missing << 'CHANNEL_ID' if CHANNEL_ID.nil? || CHANNEL_ID.empty?
  missing << 'CONTACT_ID' if CONTACT_ID.nil? || CONTACT_ID.empty?

  unless missing.empty?
    puts 'Error: Missing required environment variables:'
    missing.each { |var| puts "  - #{var}" }
    return false
  end
  true
end

def main
  return unless validate_config

  # Example 1: List all approved templates
  puts '=' * 60
  puts 'Listing approved WhatsApp templates...'
  puts '=' * 60

  begin
    templates = list_templates
    if templates.empty?
      puts 'No approved templates found.'
      puts 'Create templates in the WhatsApp Business Manager first.'
      return
    end

    puts "Found #{templates.length} template(s):\n\n"
    templates.first(5).each do |template|
      puts "  Name: #{template['name']}"
      puts "  Category: #{template['category']}"
      puts "  Language: #{template['language']}"
      puts "  Status: #{template['status']}"
      puts
    end
  rescue StandardError => e
    handle_template_error(e)
    return
  end

  # Example 2: List templates by category
  puts '=' * 60
  puts 'Listing MARKETING templates...'
  puts '=' * 60

  begin
    marketing_templates = list_templates('MARKETING')
    puts "Found #{marketing_templates.length} marketing template(s)"
  rescue StandardError => e
    handle_template_error(e)
  end

  # Example 3: Send a template with text parameters
  puts "\n#{'=' * 60}"
  puts 'Sending template with text parameters...'
  puts '=' * 60

  begin
    message = send_template_with_text_params(
      CHANNEL_ID,
      CONTACT_ID,
      'order_confirmation',
      ['John Doe', 'ORD-12345'],
      'en'
    )

    puts 'Template message sent successfully!'
    puts "  Message ID: #{message['id']}"
    puts "  Status: #{message['status']}"
  rescue StandardError => e
    handle_template_error(e)
    puts "\nNote: Update template_name to match your approved template"
  end

  # Example 4: Send template with image header
  puts "\n#{'=' * 60}"
  puts 'Sending template with image header...'
  puts '=' * 60

  begin
    message = send_template_with_header_image(
      CHANNEL_ID,
      CONTACT_ID,
      'promotion_with_image',
      'https://example.com/promo-image.jpg',
      ['Summer Sale', '50%'],
      'en'
    )

    puts 'Template with image sent successfully!'
    puts "  Message ID: #{message['id']}"
  rescue StandardError => e
    handle_template_error(e)
    puts "\nNote: Update template_name to match your approved template"
  end

  # Example 5: Send template with document header
  puts "\n#{'=' * 60}"
  puts 'Sending template with document header...'
  puts '=' * 60

  begin
    message = send_template_with_header_document(
      CHANNEL_ID,
      CONTACT_ID,
      'invoice_template',
      'https://example.com/invoice.pdf',
      'Invoice-2026-001.pdf',
      ['$199.99'],
      'en'
    )

    puts 'Template with document sent successfully!'
    puts "  Message ID: #{message['id']}"
  rescue StandardError => e
    handle_template_error(e)
    puts "\nNote: Update template_name to match your approved template"
  end
end

main if __FILE__ == $PROGRAM_NAME
